render/board.go

```diff
@@ -8,35 +8,110 @@ import (
 )
 
 const (
-	BoardSquareEmpty     = 0 // Zero State (Default)
-	BoardSquareFood      = 1
-	BoardSquareSnakeBody = 2
-	BoardSquareSnakeHead = 3
-	BoardSquareSnakeTail = 4
-	BoardSquareDeadSnake = 5
+	BoardSquareFood BoardSquareContentType = iota
+	BoardSquareSnakeBody
+	BoardSquareSnakeHead
+	BoardSquareSnakeTail
+	BoardSquareDeadSnake
+	BoardSquareHazard
 )
 
-type BoardSquareContent int
+// BoardSquareContentType works like an enum.
+// It provides a restricted set of types of content that can be placed in a board square.
+type BoardSquareContentType int
 
-type BoardSquare struct {
-	Content   BoardSquareContent
+// BoardSquareContent represents a single piece of content in a single square of the game board.
+// Examples of content are food, snake body parts and hazard squares
+type BoardSquareContent struct {
+	Type      BoardSquareContentType
 	HexColor  string
 	SnakeType string
 	Direction string
 	Corner    string
 }
 
+// BoardSquare represents a unique location on the game board.
+type BoardSquare struct {
+	Contents []BoardSquareContent
+}
+
+// Board is the root datastructure that represents a game board
 type Board struct {
 	Width   int
 	Height  int
-	Squares [][]BoardSquare
+	squares map[engine.Point]*BoardSquare
```

Comment:
I found that there were some weird behaviors/bugs that could easily happen based on how the array indexing was done. Ultimately I found it best to use a map structure for the following reasons:

- no ambiguity for how to access `x,y` coordinates (with arrays it's not always clear what should be x or y, and in this case also whether y=0 was the top or bottom of the board)
- it's a sparse data structure, we don't need to allocate space for empty squares